tests/multigrid/test_p_multigrid.py

```diff
@@ -213,3 +209,170 @@ def test_p_multigrid_mixed():
     assert solver.snes.ksp.its <= 5
     ppc = solver.snes.ksp.pc.getPythonContext().ppc
     assert ppc.getMGLevels() == 3
+
+
+def test_p_fas_scalar():
+    mat_type = "matfree"
+    mesh = UnitSquareMesh(4, 4, quadrilateral=True)
+    V = FunctionSpace(mesh, "CG", 4)
+
+    # This problem is fabricated such that the exact solution
+    # is resolved before reaching the finest level, hence no
+    # work should be done in the finest level.
+    # This will no longer be true for non-homogenous bcs, due
+    # to the way firedrake imposes the bcs before injection.
+    u = Function(V)
+    v = TestFunction(V)
+    x = SpatialCoordinate(mesh)
+    f = x[0]*(1-x[0]) + x[1]*(1-x[1])
+    bcs = DirichletBC(V, 0, "on_boundary")
+
+    F = inner(grad(u), grad(v))*dx - inner(f, v)*dx
+
+    # Due to the convoluted nature of the nested iteration
+    # it is better to specify absolute tolerances only
+    rhs = assemble(F, bcs=bcs)
+    with rhs.dat.vec_ro as Fvec:
+        Fnorm = Fvec.norm()
+
+    rtol = 1E-8
+    atol = rtol * Fnorm
+
+    coarse = {
+        "ksp_type": "preonly",
+        "ksp_norm_type": None,
+        "pc_type": "cholesky"}
+
+    relax = {
+        "ksp_type": "chebyshev",
+        "ksp_monitor_true_residual": None,
+        "ksp_norm_type": "unpreconditioned",
+        "ksp_max_it": 3,
+        "pc_type": "jacobi"}
+
+    pmg = {
+        "snes_type": "ksponly",
+        "ksp_atol": atol,
+        "ksp_rtol": 1E-50,
+        "ksp_type": "cg",
+        "ksp_converged_reason": None,
+        "ksp_monitor_true_residual": None,
+        "ksp_norm_type": "unpreconditioned",
+        "pc_type": "python",
+        "pc_python_type": "firedrake.PMGPC",
+        "pmg_pc_mg_type": "multiplicative",
+        "pmg_mg_levels": relax,
+        "pmg_mg_levels_transfer_mat_type": mat_type,
+        "pmg_mg_coarse": coarse}
+
+    pfas = {
+        "mat_type": "aij",
+        "snes_monitor": None,
+        "snes_converged_reason": None,
+        "snes_atol": atol,
+        "snes_rtol": 1E-50,
+        "snes_type": "python",
+        "snes_python_type": "firedrake.PMGSNES",
+        "pfas_snes_fas_type": "kaskade",
+        "pfas_fas_levels": pmg,
+        "pfas_fas_coarse": coarse}
+
+    problem = NonlinearVariationalProblem(F, u, bcs)
+    solver = NonlinearVariationalSolver(problem, solver_parameters=pfas)
+    solver.solve()
+
+    ppc = solver.snes.getPythonContext().ppc
+    levels = ppc.getFASLevels()
+    assert levels == 3
+    assert ppc.getFASSmoother(levels-1).getLinearSolveIterations() == 0
+
+
+def test_p_fas_nonlinear_scalar():
+    mat_type = "matfree"
+    N = 4
+    dxq = dx(degree=3*N)  # here we also test coarsening of quadrature degree
```

Comment:
Is there any way to assert in the test that the quadrature degree was indeed modified on the coarse problem?

Reply:
There must be a way. I could try to extract the forms from the coarse SNES.